src/Components/Components/src/ParameterCollectionExtensions.cs

```diff
@@ -41,17 +42,68 @@ public static class ParameterCollectionExtensions
                 _cachedWritersByType[targetType] = writers;
             }
 
+            // The logic is a little convoluted now that we have "Extra" parameters. We want to avoid allocations where
+            // possible.
+            //
+            // Error cases that are possible here:
+            // 1. Using an unknown parameter when there is no "Extra" parameter defined
+            // 2. Using an unknown parameter when there is an "Extra" parameter defined *AND* explicitly
+            //    providing a value for the extra parameter.
+            //
+            // The second case has to be an error because we want to allow users to set the "Extra" parameter
+            // explicitly, but we don't ever want to mutate a value the user gives us. We also don't want to
+            // implicitly copy a value the user gives us. Either one of those implementation choices would do
+            // something unexpected.
+
+            var isExtraParameterSetExplicitly = false;
+            Dictionary<string, object> extras = null;
```

Comment:
I know it's a bit nitpicky and unquantified-micro-optimizationy, but could we skip most of these extra per-param tests inside the loop if we know the receiver component doesn't have any "extra attributes" parameter (e.g., if `writers.CaptureExtraAttributesPropertyName == null`)?

If we checked for that once up front, could we eliminate the many of the checks inside the loop, even if it comes at the cost of a bit of extra code duplication?

It would be cool if this feature was basically zero cost in the case where you're not using it.

Reply:
I'll try this out. It might even be a lot simpler to read this way.

Reply:
Take a look now, it's vastly simplified by breaking it up into two loops 👍